### DISCUSSION TITLE
How to do double inheritance?

### DISCUSSION THREAD

**Discussion by martijn-van-den-bergh** [clientIO/joint#2193] [Q&A]
### Introduction

I am having trouble understanding why my code won't compile. I want to make a BaseShape that has functinality that all my shapes should have. Then i want to make custom shapes based on that BaseShape. But my compiler is giving me errors.

Here are my two shapes:

```
import { dia, util } from '@clientio/rappid';

export class BaseShape extends dia.Element {

    defaults() {
        return {
            ...super.defaults,
            type: 'myApp.BaseShape',
            size: {
                width: 100,
                height: 80
            },
        }
    }
    markup = [
        {
            tagName: 'rect',
            selector: 'body'
        },
        {
            tagName: 'text',
            selector: 'label'
        }
    ]

    setText(text: string) {
        const wrapText = util.breakText(text, {
            width: this.attr('body/width'),
            height: this.attr('body/height'),
        });

        this.attr('label/text', wrapText);
    }
}

export class CustomShape1 extends BaseShape {

    defaults() {
        return {
            ...super.defaults,
            type: 'myApp.CustomShape1',
        }
    }
}
```

And I get the following error:
`Property 'defaults' in type 'CustomShape1' is not assignable to the same property in base type 'BaseShape'.
  Type '() => { type: string; }' is not assignable to type '() => { type: string; size: { width: number; height: number; }; }'.
    Property 'size' is missing in type '{ type: string; }' but required in type '{ type: string; size: { width: number; height: number; }; }'.`

No i understand what happens. I should add the size property to my CustomShape1. But I don't want to do that. And in the docs it states the following:

`Using ...super.defaults, if an attribute is undefined in the child, the parent attribute will be assigned instead. Similarly, once a property is set in the child, additional values of the same property from the parent are replaced.`

In this tutorial: https://resources.jointjs.com/tutorial/ts-shape

But this error message is giving me a conflicting message. Am I understanding it wrong?

### Does your question relate to JointJS or JointJS+. Select both if applicable.

JointJS

**Comment by kumilingus** [comment]
The `defaults` is a property on `dia.Element`. It is a method on `BaseShape`.

```ts
export class CustomShape1 extends BaseShape {

    defaults() {
        return {
            ...super.defaults(), 
            type: 'myApp.CustomShape1',
        }
    }
}
```

**Reply by martijn-van-den-bergh** [reply]
ah, damn you're right. thanks!
